experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp

```diff
@@ -22,99 +22,146 @@
 
 using namespace mlir;
 
-namespace {
-struct ModuloSchedulingPass
-    : public ModuloSchedulingPassBase<ModuloSchedulingPass> {
-  // ModuloScheduling(int unrollFactor):unrollFactor_(unrollFactor){}
-  ModuloSchedulingPass() = default;
-  ModuloSchedulingPass(const ModuloSchedulingPass &pass) {}
-  void getDependentDialects(DialectRegistry &registry) const override {
-    registry.insert<scf::SCFDialect>();
-  }
-  void runOnFunction() override {
-    // Get the current FuncOp operation being operated on.
-    auto f = getFunction();
-
-    scf::ForOp loop;
-
-    //  Unroll the kernel
-    f.walk([&](scf::ForOp forop) {
-      if (forop.getNumIterOperands()) {
-        loop = forop;
-      }
-    });
-
-    if (loop) {
-      // Unroll
-      auto annotateFn = [this](unsigned i, Operation *op, OpBuilder b) {
-        op->setAttr("unrolled_iteration", b.getUI32IntegerAttr(i));
-      };
+namespace{
+  enum StageType{
+    Compute,
+    Load
+  };
 
-      (void)loopUnrollByFactor(loop, 2, annotateFn);
+  struct ModuloSchedulingPass : public ModuloSchedulingPassBase<ModuloSchedulingPass>
```

Comment:
note that there is a loop pipelining pass that does similar things, could it be adapted to your uses?

Reply:
If you refer to `ForLoopPipelining`, yes, it is exactly what I am using !